File: composeApp/src/commonMain/kotlin/org/hhp227/concafe/presentation/main/ranking/RankingUiState.kt
package org.hhp227.concafe.presentation.main.ranking

import org.hhp227.concafe.domain.model.RankingFeedEntry
import org.hhp227.concafe.domain.model.RankingPeriod
import org.hhp227.concafe.domain.model.RankingPromoAd

data class RankingUiState(
    val isLoading: Boolean = false,
    val errorMessage: String? = null,
    val selectedTab: TabType = TabType.MAIDS,
    val selectedPeriod: RankingPeriod = RankingPeriod.WEEKLY,
    val selectedRegion: RegionFilter = RegionFilter.ALL,
    val selectedAdIndex: Int = 0,
    val ads: List<RankingPromoAd> = emptyList(),
    val maidRankings: List<RankingFeedEntry> = emptyList(),
    val cafeRankings: List<RankingFeedEntry> = emptyList()
) {
    enum class TabType(val label: String) {
        MAIDS("메이드 랭킹"),
        CAFES("카페 랭킹")
    }

    enum class RegionFilter(val label: String, val country: String?, val city: String?) {
        ALL("전체", null, null),
        SEOUL("서울", "KR", "Seoul"),
        TOKYO("도쿄", "JP", "Tokyo"),
        OSAKA("오사카", "JP", "Osaka")
    }

    val currentAd: RankingPromoAd
        get() = ads.getOrElse(selectedAdIndex.coerceIn(0, (ads.size - 1).coerceAtLeast(0))) {
            RankingPromoAd("", "", "", "", "", "F6A8C5", "FFC8A2", "🎀")
        }

    val rankingEntries: List<RankingFeedEntry>
        get() = if (selectedTab == TabType.MAIDS) maidRankings else cafeRankings

    companion object {
        val empty = RankingUiState()
    }
}

File: iosApp/iosApp/Presentation/Main/Ranking/RankingUiState.swift
//
//  RankingUiState.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/06.
//

import Foundation
import Shared

struct RankingUiState {
    enum TabType: String, CaseIterable {
        case maids = "캐스트 랭킹"
        case cafes = "카페 랭킹"
    }

    enum RegionFilter: String, CaseIterable {
        case all = "전체"
        case seoul = "서울"
        case tokyo = "도쿄"
        case osaka = "오사카"

        var country: String? {
            switch self {
            case .all:
                return nil
            case .seoul:
                return "KR"
            case .tokyo, .osaka:
                return "JP"
            }
        }

        var city: String? {
            switch self {
            case .all:
                return nil
            case .seoul:
                return "Seoul"
            case .tokyo:
                return "Tokyo"
            case .osaka:
                return "Osaka"
            }
        }
    }

    var isLoading: Bool = false
    var isLoggedIn: Bool = false
    var isLoginPromptVisible: Bool = false
    var errorMessage: String? = nil
    var selectedTab: TabType = .maids
    var selectedPeriod: RankingPeriod = .weekly
    var selectedRegion: RegionFilter = .all
    var selectedAdIndex: Int = 0
    var ads: [Shared.RankingPromoAd] = []
    var maidRankings: [Shared.RankingFeedEntry] = []
    var cafeRankings: [Shared.RankingFeedEntry] = []

    var currentAd: Shared.RankingPromoAd {
        ads.isEmpty
            ? Shared.RankingPromoAd(id: "", badge: "", title: "", subtitle: "", desc: "", startColorHex: "F6A8C5", endColorHex: "FFC8A2", symbol: "🎀")
            : ads[min(max(selectedAdIndex, 0), ads.count - 1)]
    }

    var rankingEntries: [Shared.RankingFeedEntry] {
        selectedTab == .maids ? maidRankings : cafeRankings
    }

    static let empty = RankingUiState()

}
